pydantic/_internal/_std_types_schema.py

```diff
@@ -614,56 +600,185 @@ def sequence_like_prepare_pydantic_annotations(source_type: Any, annotations: It
+    input_value: Any, handler: core_schema.ValidatorFunctionWrapHandler, default_default_factory: Callable[[], Any]
+) -> collections.defaultdict[Any, Any]:
+    if isinstance(input_value, collections.defaultdict):
+        default_factory = input_value.default_factory
+        return collections.defaultdict(default_factory, handler(input_value))
+    else:
+        return collections.defaultdict(default_default_factory, handler(input_value))
+
+
+def get_defaultdict_default_default_factory(values_source_type: Any) -> Callable[[], Any]:
+    def infer_default() -> Callable[[], Any]:
+        allowed_default_types: dict[Any, Any] = {
+            typing.Tuple: tuple,
+            tuple: tuple,
+            collections.abc.Sequence: tuple,
+            collections.abc.MutableSequence: list,
+            typing.List: list,
+            list: list,
+            typing.Sequence: list,
+            typing.Set: set,
+            typing.MutableSet: set,
+            collections.abc.MutableSet: set,
+            collections.abc.Set: frozenset,
+            typing.MutableMapping: dict,
+            typing.Mapping: dict,
+            collections.abc.Mapping: dict,
+            collections.abc.MutableMapping: dict,
+            float: float,
+            int: int,
+            str: str,
+            bool: bool,
+        }
+        values_type_origin = get_origin(values_source_type) or values_source_type
+        instructions = 'set using `DefaultDict[..., Annotated[..., Field(default_factory=...)]]'
+        if isinstance(values_type_origin, TypeVar):
+
+            def type_var_default_factory() -> None:
+                raise RuntimeError(
+                    'Generic defaultdict cannot be used without a concrete value type or an'
+                    ' explicit default factory, ' + instructions
+                )
 
-    try:
-        keys_arg, values_arg = get_args(obj)
-    except ValueError:
-        # not argument bare `OrderedDict` is equivalent to `OrderedDict[Any, Any]`
-        return _ordered_dict_any_schema()
+            return type_var_default_factory
+        elif values_type_origin not in allowed_default_types:
+            # a somewhat subjective set of types that have reasonable default values
+            allowed_msg = ', '.join([t.__name__ for t in set(allowed_default_types.values())])
+            raise PydanticSchemaGenerationError(
+                f'Unable to infer a default factory for with keys of type {values_source_type}.'
+                f' Only {allowed_msg} are supported, other types require an explicit default factory'
+                ' ' + instructions
+            )
+        return allowed_default_types[values_type_origin]
 
-    if keys_arg == typing.Any and values_arg == typing.Any:
-        # `OrderedDict[Any, Any]`
-        return _ordered_dict_any_schema()
+    # Assume Annotated[..., Field(...)]
+    field_info = next((v for v in get_args(values_source_type) if isinstance(v, FieldInfo)), None)
```

Comment:
This works for now, but I feel like it’s a bit hacky. I think we should try to better integrate this later, eg to give a meaningful error if `Field(default_factory=…)` is used in places where it isn’t allowed (we currently ignore it)

Reply:
It worries me a bit to not actually check that `values_source_type` is actually annotated before iterating over the `get_args` call's results. I'm thinking we could do something like this:

https://github.com/pydantic/pydantic/blob/6100242ff4fbd97e51dd625bb90c6e7683371295/pydantic/_internal/_generate_schema.py#L217

to ensure this and make it safer here in the event that `get_args` starts handling more things in the future or whatever.

(Might make sense to make that a utility function or similar anyway.)

If you disagree, I don't feel strongly.